Create reactive fluid transport material model

naliboff opened the pull request:
A draft material model, which composites on top of existing material models for solid deformation to also simulate reactive fluid transport. This is based on the contribution from #5214 and discussions with @jdannberg, @danieldouglas92, and @ryanstoner1.

The material is a work in progress and the PR was created to facilitate discussion. At present, the only option is to simulate non-reactive (insoluble) fluid flow, but the idea is to maintain a flexible structure and allow users to use add additional petrologic models of varying complexity.

* [X] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](https://aspect-documentation.readthedocs.io/en/latest/user/extending/testing/writing-tests.html) for the new feature/benchmark in the [tests/](../blob/main/tests/) directory.
* [x] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

jdannberg commented on source/material_model/reactive_fluid_transport.cc:
```diff
@@ -0,0 +1,323 @@
+                  porosity_change = -in.composition[q][porosity_idx];
+
+                if (c == bound_fluid_idx && this->get_timestep_number() > 0)
+                  reaction_rate_out->reaction_rates[q][c] = - porosity_change / fluid_reaction_time_scale
+                                                            + in.composition[q][bound_fluid_idx] * trace(in.strain_rate[q]);
```
```suggestion
                  reaction_rate_out->reaction_rates[q][c] = - porosity_change / fluid_reaction_time_scale;
```
We discussed this today (with @bobmyhill) and agreed that this term probably should not be there.

jdannberg commented on source/material_model/reactive_fluid_transport.cc:
```diff
@@ -0,0 +1,323 @@
+      for (unsigned int q=0; q<in.temperature.size(); ++q)
+        {
+          // A fluid-rock reaction model where no reactions occur. The
+          // the melt (fluid) fraction at any point is thus equal to the
```
```suggestion
          // melt (fluid) fraction at any point is thus equal to the
```

jdannberg commented on source/material_model/reactive_fluid_transport.cc:
```diff
@@ -0,0 +1,323 @@
+      for (unsigned int q=0; q<in.temperature.size(); ++q)
+        {
+          // A fluid-rock reaction model where no reactions occur. The
+          // the melt (fluid) fraction at any point is thus equal to the
+          // porosity, which is determined by the assigned initial porosity,
+          // fluid boundary conditions, and fluid transport through the model.
```
I am not sure that this is true. If there is bound water, the total fluid fraction would not equal the porosity. 
The free fluid fraction is equal to the porosity.

naliboff replied:
Good point, I've modified the documentation and code accordingly.

jdannberg commented on source/material_model/reactive_fluid_transport.cc:
```diff
@@ -0,0 +1,323 @@
+                                   "Material model that is designed to advect fluids and compute "
+                                   "fluid release and absorption based on different models for "
+                                   "fluid-rock interaction. At present, a model where no fluid-rock "
+                                   "interactions occurs is the only available model. The properties "
```
```suggestion
                                   "interactions occur is the only available model. The properties "
```

jdannberg commented on tests/reactive_fluid_transport.prm:
```diff
@@ -0,0 +1,155 @@
+# This is a 1D test case for an insoluble fluid moving upwards in a
+# narrow channel through porous coupled two-phase flow.
```
The tester says: "test reactive_fluid_transport.prm is missing a folder with reference data"

jdannberg replied:
I think this would be more correct
```suggestion
# This is a 1D test case for a fluid moving upwards in a narrow channel 
#through porous coupled two-phase flow without reacting with the solid rock.
```

jdannberg commented on tests/reactive_fluid_transport.prm:
```diff
@@ -0,0 +1,155 @@
+subsection Initial composition model
+  set Model name = function
+  subsection Function
+    set Function expression = if (y<=2e2,0.1,0); 0;
```
Does this actually work with the semicolon at the end?
```suggestion
    set Function expression = if (y<=2e2,0.1,0); 0
```

naliboff replied:
It does run, but good catch (it should be removed)

jdannberg commented on tests/reactive_fluid_transport.prm:
```diff
@@ -0,0 +1,155 @@
+end
+
+
+# We use a custom material model that implements the layers with
+# different solubility.
```
This is outdated

jdannberg commented on tests/reactive_fluid_transport.prm:
```diff
@@ -0,0 +1,155 @@
+    set Reference fluid viscosity = 10
+    set Reference permeability = 2.5e-6
+    set Exponential fluid weakening factor = 0
+    set Fluid reaction time scale for operator splitting = 5e4
```
But operator splitting is not used?

naliboff replied:
Good catch, removed

jdannberg reviewed (commented):
Thank you! I have looked through your PR and had some smaller comments, and I also think it would be good if you could reorder the functions. Usually, all material models have the same order of the functions (for example, the last 4 are always evaluate(), declare_parameters(), parse_parameters(), and create_additional_named_outputs()). Have a look at the other melt material models for the right order.

Maybe it would also be good if someone else could have a look (just to comment on if the general idea of this makes sense since it duplicates some code).

naliboff commented:
@jdannberg - Thanks for the initial comments and suggestions, which I've addressed aside from the reordering the functions. I'll do the reordering next, finish the test case, and then add in a structure for specifying what reaction model to use. After that, a look from a second reviewer would be ideal.

naliboff commented:
@jdannberg @ryanstoner1 @danieldouglas92 - The functions have been re-ordered, a switch case is available for implementing different reaction schemes (current option is no reaction) and the test results have been added. It is probably ready for another round of reviews?

@bobmyhill - Would you be willing to take a look?

naliboff commented:
/rebuild

ryanstoner1 commented on source/material_model/reactive_fluid_transport.cc:
```diff
@@ -0,0 +1,350 @@
+              case no_reaction:
+              {
+                // A fluid-rock reaction model where no reactions occur.
+                // The melt (fluid) fraction at any point is equal
```
Perhaps specify this to make a bit more approachable? 
```suggestion
                // The melt (fluid) fraction (volume percent) at any point is equal
```

ryanstoner1 commented on source/material_model/reactive_fluid_transport.cc:
```diff
@@ -0,0 +1,350 @@
+                // porosity, fluid boundary conditions, and fluid
+                // transport through the model.
+                const unsigned int bound_fluid_idx = this->introspection().compositional_index_for_name("bound_fluid");
+                melt_fractions[q] = in.composition[q][bound_fluid_idx] + in.composition[q][porosity_idx];
```
Eventually should make the switch to generic fluid instead of melt for correctness.

ryanstoner1 commented on source/material_model/reactive_fluid_transport.cc:
```diff
@@ -0,0 +1,350 @@
+    {
+      base_model->evaluate(in,out);
+
+      // Modify the viscosity from the base model based on the presence of fluid.
+      const unsigned int porosity_idx = this->introspection().compositional_index_for_name("porosity");
+
```
```suggestion
      const unsigned int porosity_idx = this->introspection().compositional_index_for_name("porosity");
      // Modify the viscosity from the base model based on the presence of fluid.


```

ryanstoner1 commented on source/material_model/reactive_fluid_transport.cc:
```diff
@@ -0,0 +1,350 @@
+
+              if (in.requests_property(MaterialProperties::viscosity))
+                {
+                  const double phi_0 = 0.05;
+                  porosity = std::max(porosity,1e-8);
```
Comment could be nice here.

ryanstoner1 commented on source/material_model/reactive_fluid_transport.cc:
```diff
@@ -0,0 +1,350 @@
+        {
+          prm.declare_entry("Base model","visco plastic",
+                            Patterns::Selection(MaterialModel::get_valid_model_names_pattern<dim>()),
+                            "The name of a material model that will be modified by the "
```
```suggestion
                            "The name of a material model incorporating the "
```
Clarity.

ryanstoner1 reviewed (commented):
@naliboff Some comments about settling on the semantics. Take them of leave them. Perhaps we can discuss some when we talk. Also, it might be good to go over how to best make this extensible if new users have their own reaction formulation.

jdannberg commented on source/material_model/reactive_fluid_transport.cc:
```diff
@@ -0,0 +1,353 @@
+                // porosity, fluid boundary conditions, and fluid
+                // transport through the model.
+                const unsigned int bound_fluid_idx = this->introspection().compositional_index_for_name("bound_fluid");
+                melt_fractions[q] = in.composition[q][bound_fluid_idx] + in.composition[q][porosity_idx];
```
You are here assuming that your bound water fraction is a volume fraction (since the porosity is always a volume fraction). It would be good to document that somewhere (i.e. in the description of the material model). 

Technically, this also requires that your base model is incompressible (otherwise, the advection equation would only be valid for mass, not volume fractions). So maybe add an Assert somewhere.

jdannberg commented on source/material_model/reactive_fluid_transport.cc:
```diff
@@ -0,0 +1,353 @@
+                             "computed. If the model does not use operator splitting, this parameter is not used. "
+                             "Units: yr or s, depending on the ``Use years "
+                             "in output instead of seconds'' parameter.");
+          prm.declare_entry ("Fluid solid reaction scheme", "no reaction",
+                             Patterns::Selection("no reaction"),
+                             "Select what type of scheme to use for reactions between fluid and solid phases. "
+                             "The only present option is to have no reactions between the two phases.");
```
I am not really happy with having this option as long as there is not really an option, so you have to promise that you and @danieldouglas92 will get the first actual reaction scheme finalized soon :smile: 

Or you simply implement both options I mentioned above.

jdannberg commented on include/aspect/material_model/reactive_fluid_transport.h:
```diff
@@ -0,0 +1,153 @@
+
+        /**
+         * Enumeration for selecting which type of scheme to use for reactions between.
+         * fluids and solids. The only current option is a scheme where no reactions occur.
```
That is not true. Your scheme has a zero solubility, but there will be reactions if there is bound water at the start. So you either need to change the implementation (if you actually want no reactions) or the name of the input parameter and the documentation (to something like `zero_solubility`). 

Or you add both, and the you actually have a choice between the two!

jdannberg commented on include/aspect/material_model/reactive_fluid_transport.h:
```diff
@@ -0,0 +1,153 @@
+         */
+        enum ReactionScheme
+        {
+          no_reaction
```
See above

jdannberg commented on source/material_model/reactive_fluid_transport.cc:
```diff
@@ -0,0 +1,353 @@
+                             "computed. If the model does not use operator splitting, this parameter is not used. "
+                             "Units: yr or s, depending on the ``Use years "
+                             "in output instead of seconds'' parameter.");
+          prm.declare_entry ("Fluid solid reaction scheme", "no reaction",
```
I think I would prefer this with a dash
```suggestion
          prm.declare_entry ("Fluid-solid reaction scheme", "no reaction",
```

jdannberg commented on tests/reactive_fluid_transport_no_reaction/screen-output:
```diff
@@ -0,0 +1,79 @@
+
+
+   Postprocessing:
+     Compositions min/max/mass:   -5.788e-07/0.11/5159 // 0/0/0
```
Are you sure this is working as intended? Shouldn't the fluid be released, i.e. the porosity should increase?

naliboff replied:
Yep, good catch! The initial assignment of values (porosity, bound_fluid) was reversed.

jdannberg commented on source/material_model/reactive_fluid_transport.cc:
```diff
@@ -0,0 +1,353 @@
+          std::vector<double> eq_free_fluid_fractions(out.n_evaluation_points());
+          melt_fractions(in, eq_free_fluid_fractions);
+
+          for (unsigned int q=0; q<out.n_evaluation_points(); ++q)
+            for (unsigned int c=0; c<in.composition[q].size(); ++c)
+              {
+                double porosity_change = eq_free_fluid_fractions[q] - in.composition[q][porosity_idx];
+                // do not allow negative porosity
+                if (in.composition[q][porosity_idx] + porosity_change < 0)
+                  porosity_change = -in.composition[q][porosity_idx];
+
+                if (c == bound_fluid_idx && this->get_timestep_number() > 0)
+                  reaction_rate_out->reaction_rates[q][c] = - porosity_change / fluid_reaction_time_scale;
+                else if (c == porosity_idx && this->get_timestep_number() > 0)
+                  reaction_rate_out->reaction_rates[q][c] = porosity_change / fluid_reaction_time_scale;
+                else
+                  reaction_rate_out->reaction_rates[q][c] = 0.0;
+              }
```
This only works if you use operator splitting in your model. So (1), please document that, and (2) change your test to actually use operator splitting (because at the moment, it does not test this implementation because it does not use the reaction rates).

naliboff replied:
Can you elaborate on what part of the code needs additional documentation for the operator splitting? There is some documentation at the top on this above the following if statement (`if (this->get_parameters().use_operator_splitting && reaction_rate_out != nullptr)`), but happy to more in an additional location.

The test has been updated

jdannberg replied:
Just having the assert is fine!

jdannberg commented on tests/reactive_fluid_transport_no_reaction.prm:
```diff
@@ -0,0 +1,142 @@
+set End time                               = 1e1
+
+
+set Use operator splitting                     = false
```
You need to use operator splitting for your reactions to work.

jdannberg reviewed (changes requested):
Thank you! The structure looks good now, but I think the implementation requires some changes. Most importantly, think about if you actually want no reactions, a zero solubility, or both.

naliboff commented:
@jdannberg - Thanks for the thorough review! I've hopefully addressed most of the comments, and I think this is ready for another review.

jdannberg commented on tests/reactive_fluid_transport_no_reaction.prm:
```diff
@@ -0,0 +1,144 @@
+    set Reference fluid viscosity = 10
+    set Reference permeability = 2.5e-6
+    set Exponential fluid weakening factor = 0
+    set Fluid-solid reaction scheme = zero solubility
```
Shouldn't this be no reaction? 
You switch the operator splitting off, so technically, this should give you the correct result, but I find it to be confusing.

jdannberg replied:
I see it actually does not work because
`Exception 'ExcMessage("The Fluid-reaction scheme zero solubility must be used with operator splitting.")'`
:-)

jdannberg commented on tests/reactive_fluid_transport_zero_solubility.prm:
```diff
@@ -0,0 +1,161 @@
+
```
Can you remove this empty line?

jdannberg commented on tests/reactive_fluid_transport_no_reaction.prm:
```diff
@@ -0,0 +1,144 @@
+# This model tests the reactive fluid transport material
+# 'no reaction' scenario to ensure that there is no reaction
+# between the fluid and solid phases. The values of
+# porosity (0) and the bound fluid (0.01) are initially
+# spatially uniform. The absence of reactions between
+# the two phases and near zero velocities (free-slip on
+# all sides) produces no change in their values through time.
```
This is very similar to the zero solubility test below. Can you include the other test here with
include $ASPECT_SOURCE_DIR/tests/reactive_fluid_transport_zero_solubility.prm
and the only list the parameters that are different? This way you do not have to duplicate everything.

jdannberg commented on tests/reactive_fluid_transport_zero_solubility.prm:
```diff
@@ -0,0 +1,161 @@
+subsection Solver parameters
+  subsection Operator splitting parameters
+    # We choose the size of the reaction time step as 100 years, small enough
+    # so that it can accurately model melting and freezinddg.
```
```suggestion
    # so that it can accurately model melting and freezing.
```

jdannberg commented on tests/reactive_fluid_transport_zero_solubility.prm:
```diff
@@ -0,0 +1,161 @@
+# we use the operator splitting scheme.
+set Use operator splitting                     = true
+
+# Mellt transport is turned on to ensure it works
```
```suggestion
# Melt transport is turned on to ensure it works
```

jdannberg commented on tests/reactive_fluid_transport_zero_solubility.prm:
```diff
@@ -0,0 +1,161 @@
+
+subsection Solver parameters
+  subsection Operator splitting parameters
+    # We choose the size of the reaction time step as 100 years, small enough
```
```suggestion
    # We choose the size of the reaction time step as 1000 years, small enough
```

jdannberg commented on tests/reactive_fluid_transport_zero_solubility/screen-output:
```diff
@@ -0,0 +1,41 @@
+
+
+   Postprocessing:
+     Compositions min/max/mass:   0.01/0.01/4e+04 // 0/0/0
+     Melt fraction min/total/max: 0.01, 4e+04, 0.01
```
This shows nicely that it worked!

jdannberg commented on source/material_model/reactive_fluid_transport.cc:
```diff
@@ -0,0 +1,376 @@
+            {
+              AssertThrow(fluid_reaction_time_scale >= this->get_parameters().reaction_time_step,
+                          ExcMessage("The reaction time step " + Utilities::to_string(this->get_parameters().reaction_time_step)
+                                     + " in the operator splitting scheme is too large to compute melting rates! "
```
```suggestion
                                     + " in the operator splitting scheme is too large to compute fluid release rates! "
```

jdannberg commented on source/material_model/reactive_fluid_transport.cc:
```diff
@@ -0,0 +1,376 @@
+              AssertThrow(fluid_reaction_time_scale >= this->get_parameters().reaction_time_step,
+                          ExcMessage("The reaction time step " + Utilities::to_string(this->get_parameters().reaction_time_step)
+                                     + " in the operator splitting scheme is too large to compute melting rates! "
+                                     "You have to choose it in such a way that it is smaller than the 'Melting time scale for "
```
```suggestion
                                     "You have to choose it in such a way that it is smaller than the 'Fluid reaction time scale for "
```

jdannberg commented on source/material_model/reactive_fluid_transport.cc:
```diff
@@ -0,0 +1,376 @@
+            {
+              AssertThrow(this->get_parameters().use_operator_splitting,
+                          ExcMessage("The Fluid-reaction scheme zero solubility must be used with operator splitting."));
+              AssertThrow(this->get_material_model().is_compressible() == false,
+                          ExcMessage("The Fluid-reaction scheme zero solubility must be used with an incompressible base model."));
```
This does not work well for all models, since at this point we have not parsed the parameters from the base model. So if the base model has a compressibility that depends on an input parameter, at this point in the code we will not know about it. It would be better to move this after parsing the parameters from the base model or ito the initialize() function.

jdannberg reviewed (commented):
I only have a few minor comments, otherwise this looks good!

naliboff commented:
@jdannberg - Thanks for the last round of edits! I addressed all of the comments, but there is still a bug I am tracking down so not ready to merge yet.

naliboff commented:
@jdannberg - I think this is ready to merged!

jdannberg commented:
This looks good to me now! The only other comment I have is that this also does not create the additional material model outputs the base model has (which I fixed for the prescribed viscosity and depth dependent model here https://github.com/geodynamics/aspect/pull/5547). 

But that could be addressed in a follow-up PR. 

And sorry this took me so long to look at!
